File: src/lexer.c
/*  
    Copyright (c) MURANGWA Pacifique. and affiliates.
    This source code is licensed under the Apache License 2.0 found in the
    LICENSE file in the root directory of this source tree.
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>

#include "headers/common.h"
#include "headers/lexer.h"
#include "headers/exit-codes.h"
#include "headers/errors.h"
#include "headers/error-codes.h"



char* source; /* Source code */
int currentPos = 0; /* Current position in source */
int line = 1; /* Current line number */ 

/* Function to initialize the source code */
void initLexersSource() {
    source = source_code_info.buffer;
}

/* Function to create a new token with a lexeme */
Token makeTokenWithLexeme(TokenType type, char* lexeme) {
    Token token;
    token.type = type;
    token.lexeme = lexeme;
    token.line = line;
    return token;
}

/* Function to get the current character */
char getCurrentChar() {
  return source[currentPos];
}

/*Function to get current character and advance*/
char consume() {
  char c = getCurrentChar();
  advance(&currentPos);
  return c;
}

/* Function to scan an identifier or keyword */
Token scanIdentifierOrKeyword() {
    int start = currentPos;

    while (isalnum(getCurrentChar()) || getCurrentChar() == '_') {
      advance(&currentPos);
    }

    char* lexeme = strndup(&source[start], currentPos - start);

    /* Check if it's a keyword */
    if (strcmp(lexeme, "ubusa") == 0) return makeTokenWithLexeme(TOKEN_UBUSA, lexeme);
    if (strcmp(lexeme, "niba") == 0) return makeTokenWithLexeme(TOKEN_NIBA, lexeme);
    if (strcmp(lexeme, "nibyo") == 0) return makeTokenWithLexeme(TOKEN_NIBYO, lexeme);
    if (strcmp(lexeme, "sibyo") == 0) return makeTokenWithLexeme(TOKEN_SIBYO, lexeme);
    if (strcmp(lexeme, "nanone_niba") == 0) return makeTokenWithLexeme(TOKEN_NANONE_NIBA, lexeme);
    if (strcmp(lexeme, "umubare") == 0) return makeTokenWithLexeme(TOKEN_UMUBARE, lexeme);
    if (strcmp(lexeme, "umubare_wibice") == 0) return makeTokenWithLexeme(TOKEN_UMUBARE_WIBICE, lexeme);
    if (strcmp(lexeme, "niba_byanze") == 0) return makeTokenWithLexeme(TOKEN_NIBA_BYANZE, lexeme);
    if (strcmp(lexeme, "subiramo") == 0) return makeTokenWithLexeme(TOKEN_SUBIRAMO, lexeme);
    if (strcmp(lexeme, "tanga") == 0) return makeTokenWithLexeme(TOKEN_TANGA, lexeme);
    if (strcmp(lexeme, "porogaramu_ntoya") == 0) return makeTokenWithLexeme(TOKEN_POROGARAMU_NTOYA, lexeme);
    if (strcmp(lexeme, "tangaza_amakuru") == 0) return makeTokenWithLexeme(TOKEN_TANGAZA_AMAKURU, lexeme);
    if (strcmp(lexeme, "injiza_amakuru") == 0) return makeTokenWithLexeme(TOKEN_INJIZA_AMAKURU, lexeme);
    if (strcmp(lexeme, "komeza") == 0) return makeTokenWithLexeme(TOKEN_KOMEZA, lexeme);
    if (strcmp(lexeme, "hagarara") == 0) return makeTokenWithLexeme(TOKEN_HAGARARA, lexeme);
    if (strcmp(lexeme, "ubwoko") == 0) return makeTokenWithLexeme(TOKEN_UBWOKO, lexeme);
    if (strcmp(lexeme, "reka") == 0) return makeTokenWithLexeme(TOKEN_REKA, lexeme);
    if (strcmp(lexeme, "soma_inyandiko") ==0 ) return makeTokenWithLexeme(TOKEN_SOMA_INYANDIKO, lexeme);
    if (strcmp(lexeme, "andika_inyandiko") == 0) return makeTokenWithLexeme(TOKEN_ANDIKA_INYANDIKO, lexeme);
    if (strcmp(lexeme, "vugurura_inyandiko") == 0) return makeTokenWithLexeme(TOKEN_KUVUGURURA_INYANDIKO, lexeme);
    if (strcmp(lexeme, "kin_hagarara") == 0) return makeTokenWithLexeme(TOKEN_KIN_HAGARARA, lexeme);
    if (strcmp(lexeme, "sisitemu") == 0) return makeTokenWithLexeme(TOKEN_SISITEMU, lexeme);

    /* Not a keyword, it's an identifier */
    return makeTokenWithLexeme(TOKEN_IDENTIFIER, lexeme);
}

/* Function to scan a string literal */
Token scanStringLiteral() {
    char delimiter = consume(); /* Single or double quote */
    int start = currentPos ; /* get position where we have delimeter */

    while (getCurrentChar() != delimiter) {
        if (getCurrentChar() == '\0' || getCurrentChar() == '\n') {
            syntaxError(ERROR_UNTERMINATED_STRING_LITERAL, '\n', line);
        }
        advance(&currentPos);
    }

    char* lexeme = strndup(&source[start], currentPos - start);
    advance(&currentPos); /* escape closing delimiter for string litelar. */
    return makeTokenWithLexeme(TOKEN_STRING, lexeme);
}


/* Function to scan a number */
Token scanNumber() {
    int start = currentPos;

    while (isdigit(getCurrentChar())) {
        advance(&currentPos);
    }

    if (getCurrentChar() == '.' && isdigit(source[currentPos + 1])) {
        advance(&currentPos);
        while (isdigit(getCurrentChar())) {
            advance(&currentPos);
        }
        return makeTokenWithLexeme(TOKEN_FLOAT, strndup(&source[start], currentPos - start));
    }

    return makeTokenWithLexeme(TOKEN_NUMBER, strndup(&source[start], currentPos - start));
}

/* Function to skip whitespace and comments */
void skipWhitespaceAndComments() {
    while (true) {
        char c = getCurrentChar();
        if (c == ' ' || c == '\t' || c == '\r') {
            advance(&currentPos);
        } else if (c == '\n') {
            /* Newline character */
            advance(&currentPos);
            line++;
        } else if (c == '#') {
            /* Comment, skip until the end of the line */
            while (getCurrentChar() != '\n' && getCurrentChar() != '\0') {
              advance(&currentPos);
            }
        } else {
            break;
        }
    }
}

/* Function to scan the next token */
Token scanToken() {
    skipWhitespaceAndComments();
    char c = getCurrentChar();

    if (c == '\0') {
        /* End of file */
        return makeTokenWithLexeme(TOKEN_EOF, "EOF");
    }

    switch (c) {
        /* One-character tokens */
        case '-':
            advance(&currentPos);
            if (getCurrentChar() == '-') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_DECREMENT, "decrement");
            }
            return makeTokenWithLexeme(TOKEN_MINUS, "minus");
        case '+':
            advance(&currentPos);
            if ( getCurrentChar() == '+') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_INCREMENT, "increment");
            }
            return makeTokenWithLexeme(TOKEN_PLUS, "plus");
        case '*':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_STAR, "star");
        case '=':
            advance(&currentPos);
            if(getCurrentChar() == '='){
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_EQUALITY, "equality");
            }
            return makeTokenWithLexeme(TOKEN_ASSIGNMENT, "assignment");
        case '/':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_DIVISION, "division");
        case '^':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_POWER, "power");
        case '%':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_MODULO, "modulo");
        case '&':
            advance(&currentPos);
            if (getCurrentChar() == '&') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_AND, "and");
            }
            return makeTokenWithLexeme(TOKEN_AMPERSAND, "ampersand");
        case '|':
            advance(&currentPos);
            if (getCurrentChar() == '|') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_OR, "or");
            }
        case '!':
            advance(&currentPos);
            if (getCurrentChar() == '=') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_NOT_EQUAL, "not_equal");
            }
            return makeTokenWithLexeme(TOKEN_NEGATION, "negation");
        case ';':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_END_OF_LINE, "end-of-line");
        case ']':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_CLOSE_BRACKET, "close-bracket");
        case '[':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_OPEN_BRACKET, "open-bracket");
        case '(':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_OPEN_PARANTHESES, "open-parantheses");
        case ')':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_CLOSE_PARANTHESES, "close-parantheses");
        case '{':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_OPEN_CURLY_BRACES, "open-curly-braces");
        case '}':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_CLOSE_CURLY_BRACES, "close-curly-braces");
        case '\'':
            return scanStringLiteral();
        case '"':
            return scanStringLiteral();
        case ':':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_COLON, "colon");
        case '`':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_BACK_TICKS, "back-ticks");
        case '>':
            advance(&currentPos);
            if (getCurrentChar() == '=') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_GREATER_THAN_OR_EQUAL, "greater-than-or-equal");
            }
            return makeTokenWithLexeme(TOKEN_GREATER_THAN, "greater-than");
        case '<':
            advance(&currentPos);
            if (getCurrentChar() == '=') {
                advance(&currentPos);
                return makeTokenWithLexeme(TOKEN_LESS_THAN_OR_EQUAL, "less-than-or-equal");
            }
            return makeTokenWithLexeme(TOKEN_LESS_THAN, "less-than");
        case '.':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_PERIOD, "period");
        case ',':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_COMMA, "comma");
        case '$':
            advance(&currentPos);
            return makeTokenWithLexeme(TOKEN_DOLLA_SIGN, "dolla-sign");
        default:
            if (isdigit(c)) {
                return scanNumber();
            } else if (isalpha(c) || c == '_') {
                return scanIdentifierOrKeyword();
            } else {
                syntaxError(ERROR_UNEXPECTED_CHARACTER, c, line);
            }
    }
}

File: src/lexer/lexer.ts
/******************************************
 *                 Lexer                  *
 *     Produce tokens from the source     *
 ******************************************/

import TokenType from './tokens';
import { createKinError } from '../lib/errors';
import { Span } from '../lib/span';

/* Token structure: full source span plus legacy line field. */
export interface Token {
  type: TokenType;
  lexeme: string;
  line: number;
  column: number;
  start: number;
  end: number;
}

export function tokenSpan(token: Token): Span {
  return {
    start: token.start,
    end: token.end,
    line: token.line,
    column: token.column,
  };
}

class Lexer {
  private sourceCodes: string;
  private currentPos: number = 0;
  private currentLine: number = 1;
  private currentColumn: number = 1;

  constructor(sourceCodes: string) {
    this.sourceCodes = sourceCodes;
  }

  /* Advance one character, tracking line and column. */
  private advance(): void {
    const ch = this.sourceCodes[this.currentPos];
    this.currentPos++;
    if (ch === '\n') {
      this.currentLine++;
      this.currentColumn = 1;
    } else {
      this.currentColumn++;
    }
  }

  private peek(): string {
    return this.sourceCodes[this.currentPos];
  }

  private consume(): string {
    const char: string = this.peek();
    this.advance();
    return char;
  }

  private makeToken(
    type: TokenType,
    lexeme: string,
    start: number,
    line: number,
    column: number,
  ): Token {
    return {
      type,
      lexeme,
      line,
      column,
      start,
      end: this.currentPos,
    };
  }

  private isSingleAlphaCharacter(s: string): boolean {
    return /^[a-zA-Z]$/.test(s);
  }

  private isDigit(s: string): boolean {
    return /^[0-9]$/.test(s);
  }

  private alphaNumeric(s: string): boolean {
    return this.isDigit(s) || this.isSingleAlphaCharacter(s);
  }

  private skipWhitespaceAndComments(): void {
    while (true) {
      const c: string = this.peek();
      if (c === ' ' || c === '\t' || c === '\r') {
        this.advance();
      } else if (c === '\n') {
        this.advance();
      } else if (c === '#') {
        while (
          this.peek() !== '\n' &&
          this.sourceCodes.length !== this.currentPos
        ) {
          this.advance();
        }
      } else {
        break;
      }
    }
  }

  /**
   * Scan a non-negative number. Unary minus is a separate MINUS token
   * so that `x -5` parses as subtraction, not as `x` followed by `-5`.
   */
  private scanNumber(): Token {
    const start = this.currentPos;
    const line = this.currentLine;
    const column = this.currentColumn;

    while (this.isDigit(this.peek())) {
      this.advance();
    }
    if (
      this.peek() == '.' &&
      this.isDigit(this.sourceCodes[this.currentPos + 1])
    ) {
      this.advance();
      while (this.isDigit(this.peek())) {
        this.advance();
      }
      const nbr = this.sourceCodes.slice(start, this.currentPos);
      return this.makeToken(TokenType.FLOAT, nbr, start, line, column);
    }

    const nbr = this.sourceCodes.slice(start, this.currentPos);
    return this.makeToken(TokenType.INTEGER, nbr, start, line, column);
  }

  private scanStringLiteral(): Token {
    const start = this.currentPos;
    const line = this.currentLine;
    const column = this.currentColumn;
    const quote: string = this.consume();
    while (this.peek() !== quote) {
      if (this.peek() === '\n' || this.currentPos === this.sourceCodes.length) {
        throw createKinError('K003', {
          span: {
            start,
            end: this.currentPos,
            line,
            column,
          },
          message: `Unterminated string literal at line ${line}`,
        });
      }
      this.advance();
    }

    this.advance();
    return this.makeToken(
      TokenType.STRING,
      this.sourceCodes.slice(start + 1, this.currentPos - 1),
      start,
      line,
      column,
    );
  }

  private scanIdentifierOrKeyword(): Token {
    const start = this.currentPos;
    const line = this.currentLine;
    const column = this.currentColumn;
    while (this.alphaNumeric(this.peek()) || this.peek() === '_') {
      this.advance();
    }

    const lexeme: string = this.sourceCodes.slice(start, this.currentPos);
    const keywordType = this.keywordType(lexeme);
    if (keywordType !== undefined) {
      return this.makeToken(keywordType, lexeme, start, line, column);
    }
    return this.makeToken(TokenType.IDENTIFIER, lexeme, start, line, column);
  }

  private keywordType(lexeme: string): TokenType | undefined {
    switch (lexeme) {
      case 'niba':
        return TokenType.NIBA;
      case 'nanone_niba':
        return TokenType.NANONE_NIBA;
      case 'niba_byanze':
        return TokenType.NIBA_BYANZE;
      case 'subiramo_niba':
        return TokenType.SUBIRAMO_NIBA;
      case 'hagarara':
        return TokenType.HAGARARA;
      case 'komeza':
        return TokenType.KOMEZA;
      case 'tanga':
        return TokenType.TANGA;
      case 'porogaramu_ntoya':
        return TokenType.POROGARAMU_NTOYA;
      case 'reka':
        return TokenType.REKA;
      case 'ntahinduka':
        return TokenType.NTAHINDUKA;
      case 'gereranya':
        return TokenType.GERERANYA;
      case 'usanze':
        return TokenType.USANZE;
      case 'ibindi':
        return TokenType.IBINDI;
      case 'imiterere':
        return TokenType.IMITERERE;
      case 'tegura':
        return TokenType.TEGURA;
      case 'rema':
        return TokenType.REMA;
      case 'ikomoka':
        return TokenType.IKOMOKA;
      case 'rusange':
        return TokenType.RUSANGE;
      case 'bwite':
        return TokenType.BWITE;
      case 'ubwoko':
        return TokenType.UBWOKO;
      // Bare `_` is the current instance; `_izina` stays an identifier.
      case '_':
        return TokenType.THIS;
      default:
        return undefined;
    }
  }

  private scanToken(): Token {
    this.skipWhitespaceAndComments();

    if (this.currentPos == this.sourceCodes.length) {
      return this.makeToken(
        TokenType.EOF,
        'EOF',
        this.currentPos,
        this.currentLine,
        this.currentColumn,
      );
    }

    const start = this.currentPos;
    const line = this.currentLine;
    const column = this.currentColumn;
    const char = this.peek();

    switch (char) {
      case '-':
        this.advance();
        if (this.peek() == '-') {
          this.advance();
          return this.makeToken(TokenType.DECREMENT, '--', start, line, column);
        }
        // No longer fold digits into a negative literal; unary minus is
        // handled in the parser so `x -5` is subtraction.
        return this.makeToken(TokenType.MINUS, '-', start, line, column);
      case '+':
        this.advance();
        if (this.peek() == '+') {
          this.advance();
          return this.makeToken(TokenType.INCREMENT, '++', start, line, column);
        }
        return this.makeToken(TokenType.PLUS, '+', start, line, column);
      case '*':
        this.advance();
        return this.makeToken(TokenType.STAR, '*', start, line, column);
      case '=':
        this.advance();
        if (this.peek() == '=') {
          this.advance();
          return this.makeToken(TokenType.EQUALITY, '==', start, line, column);
        }
        return this.makeToken(TokenType.EQUAL, '=', start, line, column);
      case '/':
        this.advance();
        return this.makeToken(TokenType.DIVISION, '/', start, line, column);
      case '^':
        this.advance();
        return this.makeToken(TokenType.EXPONENT, '^', start, line, column);
      case '%':
        this.advance();
        return this.makeToken(TokenType.MODULO, '%', start, line, column);
      case '&':
        this.advance();
        if (this.peek() == '&') {
          this.advance();
          return this.makeToken(TokenType.AND, '&&', start, line, column);
        }
        return this.makeToken(TokenType.AMPERSAND, '&', start, line, column);
      case '!':
        this.advance();
        if (this.peek() == '=') {
          this.advance();
          return this.makeToken(TokenType.NOT_EQUAL, '!=', start, line, column);
        }
        return this.makeToken(TokenType.NEGATION, '!', start, line, column);
      case '|':
        this.advance();
        if (this.peek() == '|') {
          this.advance();
          return this.makeToken(TokenType.OR, '||', start, line, column);
        }
        throw createKinError('K004', {
          span: { start, end: this.currentPos, line, column },
          params: { char: '|' },
          message: `Unexpected character '|' at line ${line}`,
        });
      case ';':
        this.advance();
        return this.makeToken(TokenType.SEMI_COLON, ';', start, line, column);
      case ']':
        this.advance();
        return this.makeToken(
          TokenType.CLOSE_BRACKET,
          ']',
          start,
          line,
          column,
        );
      case '[':
        this.advance();
        return this.makeToken(TokenType.OPEN_BRACKET, '[', start, line, column);
      case '(':
        this.advance();
        return this.makeToken(
          TokenType.OPEN_PARANTHESES,
          '(',
          start,
          line,
          column,
        );
      case ')':
        this.advance();
        return this.makeToken(
          TokenType.CLOSE_PARANTHESES,
          ')',
          start,
          line,
          column,
        );
      case '{':
        this.advance();
        return this.makeToken(
          TokenType.OPEN_CURLY_BRACES,
          '{',
          start,
          line,
          column,
        );
      case '}':
        this.advance();
        return this.makeToken(
          TokenType.CLOSE_CURLY_BRACES,
          '}',
          start,
          line,
          column,
        );
      case '"':
        return this.scanStringLiteral();
      case ':':
        this.advance();
        return this.makeToken(TokenType.COLON, ':', start, line, column);
      case '>':
        this.advance();
        if (this.peek() == '=') {
          this.advance();
          return this.makeToken(
            TokenType.GREATER_THAN_OR_EQUAL,
            '>=',
            start,
            line,
            column,
          );
        }
        return this.makeToken(TokenType.GREATER_THAN, '>', start, line, column);
      case '<':
        this.advance();
        if (this.peek() == '=') {
          this.advance();
          return this.makeToken(
            TokenType.LESS_THAN_OR_EQUAL,
            '<=',
            start,
            line,
            column,
          );
        }
        return this.makeToken(TokenType.LESS_THAN, '<', start, line, column);
      case ',':
        this.advance();
        return this.makeToken(TokenType.COMMA, ',', start, line, column);
      case '.':
        this.advance();
        return this.makeToken(TokenType.DOT, '.', start, line, column);
      default:
        if (!Number.isNaN(Number(char))) {
          return this.scanNumber();
        } else if (this.isSingleAlphaCharacter(char) || char === '_') {
          return this.scanIdentifierOrKeyword();
        } else {
          throw createKinError('K004', {
            span: {
              start,
              end: start + 1,
              line,
              column,
            },
            params: { char },
            message: `Unexpected character '${char}' at line ${line}`,
          });
        }
    }
  }

  public tokenize(): Token[] {
    const tokens: Token[] = new Array<Token>();
    for (;;) {
      const token: Token = this.scanToken();
      tokens.push(token);
      if (token.type === TokenType.EOF) break;
    }
    return tokens;
  }
}

export default Lexer;
